### PR TITLE
groupfree3d transformer module

### PR DESCRIPTION
**Pull request by hjin2902**
support Group-Free 3D transformer

### CODE REVIEW

File: mmdet3d/models/model_utils/transformer.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    Args:
+        embed_dims (int): The embedding dimension.
+        num_heads (int): Parallel attention heads. Same as
+            `nn.MultiheadAttention`.
+        dropout (float):w A Dropout layer on attn_output_weights. Default: 0..
+        init_cfg (obj:`mmcv.ConfigDict`): The Config for initialization.
+            Default: None.
+    """
+
+    def __init__(self,
```
Comment (jshilong) [inline mmdet3d/models/model_utils/transformer.py:27]: Seems can use `__init__` of MultiheadAttention directly

File: mmdet3d/models/model_utils/transformer.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                 init_cfg=None,
+                 **kwargs):
+        super(GroupFree3DMultiheadAttention, self).__init__(
+            embed_dims=embed_dims,
+            num_heads=num_heads,
+            dropout=dropout,
+            init_cfg=init_cfg,
+            **kwargs)
+
+    def forward(self,
```
Comment (jshilong) [inline mmdet3d/models/model_utils/transformer.py:40]:
```
                query,
                key,
                value,
                residual,
                query_pos=None,
                key_pos=None,
                **kwargs
``` 
may be better.

File: mmdet3d/models/model_utils/transformer.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+            obj:`mmcv.ConfigDict`): Config of transformerlayer
+            in TransformerCoder. If it is obj:`mmcv.ConfigDict`,
+             it would be repeated `num_layer` times to a
+             list[`mmcv.ConfigDict`]. Default: None.
+        num_layers (int): The number of `TransformerLayer`. Default: None.
+        init_cfg (obj:`mmcv.ConfigDict`): The Config for initialization.
+            Default: None.
+    """
+
+    def __init__(self,
```
Comment (jshilong) [inline mmdet3d/models/model_utils/transformer.py:146]:
```
                 bbox_coder,
                 pred_layer_cfg=None,
                 **kwargs
``` may be better.

File: mmdet3d/models/model_utils/transformer.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                PositionEmbeddingLearned(6, self.embed_dims))
+        # key position embed
+        self.decoder_cross_posembeds = nn.ModuleList()
+        for _ in range(self.num_layers):
+            self.decoder_cross_posembeds.append(
+                PositionEmbeddingLearned(3, self.embed_dims))
+
+    def init_weights(self):
+        # follow the official Group-Free-3D to init parameters
+        for m in self.modules():
```
Comment (jshilong) [inline mmdet3d/models/model_utils/transformer.py:189]: make sure this can initialize all parameters,  please check that all parameters with dim>1 could be initialized in this loop.

File: mmdet3d/models/model_utils/transformer.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        return self.num_classes + 1
+
+    def _get_reg_out_channels(self):
+        """Return the channel number of regression outputs."""
+        # Objectness scores (1), center residual (3),
+        # heading class+residual (num_dir_bins*2),
+        # size class+residual(num_sizes*4)
+        return 3 + self.num_dir_bins * 2 + self.num_sizes * 4
+
+    def forward(self,
```
Comment (jshilong) [inline mmdet3d/models/model_utils/transformer.py:206]:
```
                candidate_features,
                candidate_xyz,
                seed_features,
                seed_xyz,
                base_bbox3d,
                **kwargs
``` 
may be better.

File: mmdet3d/models/model_utils/transformer.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            value=value,
+            residual=residual,
+            query_pos=query_pos,
+            key_pos=key_pos,
+            attn_mask=attn_mask,
+            key_padding_mask=key_padding_mask,
+            **kwargs)
+
+
+class PositionEmbeddingLearned(nn.Module):
```
Comment (ZwwWayne) [inline mmdet3d/models/model_utils/transformer.py:109]: How about `LearnedPositionalEmbedding`? If it duplicates that in mmcv/mmdet, we can use `ConvBNPositionalEmbedding`.

File: mmdet3d/models/model_utils/transformer.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            nn.Conv1d(num_pos_feats, num_pos_feats, kernel_size=1))
+
+    def forward(self, xyz):
+        xyz = xyz.transpose(1, 2).contiguous()
+        position_embedding = self.position_embedding_head(xyz)
+        return position_embedding
+
+
+@TRANSFORMER_LAYER_SEQUENCE.register_module()
+class GroupFree3DTransformerDecoder(TransformerLayerSequence):
```
Comment (ZwwWayne) [inline mmdet3d/models/model_utils/transformer.py:126]: It essentially implements a head, we can directly write a head for that.
